File: tools/autonomy/recommendOptimizationStrategy.js
/**
 * Tool: Recommend Optimization Strategy
 * Uses AI to recommend optimization strategies based on MSE trends and execution history
 */
export default {
  name: 'recommendOptimizationStrategy',
  description: 'Analyze MSE trends and execution history to recommend the best optimization strategy',
  parameters: { 
    mseHistory: 'array', // Array of MSE values from recent iterations
    currentMSE: 'number',
    targetMSE: 'number',
    iterationNumber: 'number',
    anthropic: 'object|null'
  },
  
  async execute(params, context) {
    const { mseHistory, currentMSE, targetMSE, iterationNumber, anthropic } = params;
    
    if (!anthropic) {
      // Fallback to rule-based strategy
      return this._fallbackStrategy(mseHistory, currentMSE, targetMSE, iterationNumber);
    }
    
    // Analyze trends
    const trends = this._analyzeTrends(mseHistory);
    
    // Get memory context
    const recentAttempts = await this._getRecentAttempts(context);
    
    // Use AI to recommend strategy
    const recommendation = await this._getAIRecommendation(
      { mseHistory, currentMSE, targetMSE, iterationNumber, trends, recentAttempts },
      anthropic
    );
    
    return {
      recommendation,
      trends,
      context: {
        recentAttempts: recentAttempts.length,
        currentProgress: ((targetMSE / currentMSE) * 100).toFixed(1) + '%',
        iterationNumber
      },
      timestamp: new Date().toISOString()
    };
  },
  
  /**
   * Analyze MSE trends
   */
  _analyzeTrends(mseHistory) {
    if (!mseHistory || mseHistory.length < 2) {
      return {
        trend: 'insufficient_data',
        improvement: false,
        stagnant: false
      };
    }
    
    const recent = mseHistory.slice(-5);
    const improvements = [];
    
    for (let i = 1; i < recent.length; i++) {
      improvements.push(recent[i-1] - recent[i]);
    }
    
    const avgImprovement = improvements.reduce((a, b) => a + b, 0) / improvements.length;
    const isImproving = avgImprovement > 0;
    const isStagnant = Math.abs(avgImprovement) < 0.001;
    
    return {
      trend: isImproving ? 'improving' : 'declining',
      improvement: isImproving,
      stagnant: isStagnant,
      avgImprovement,
      recentMSEs: recent
    };
  },
  
  /**
   * Get recent optimization attempts from memory
   */
  async _getRecentAttempts(context) {
    const entries = context.memory.entries || [];
    return entries.filter(e => 
      e.key && e.key.startsWith('iteration_') && 
      e.value && typeof e.value.mse === 'number'
    ).slice(-10);
  },
  
  /**
   * Use AI to recommend optimization strategy
   */
  async _getAIRecommendation(analysisContext, anthropic) {
    const systemPrompt = `You are an expert ML optimization strategist specializing in cryptocurrency price prediction.
Your role is to analyze MSE trends and recommend the most effective optimization strategy.

Consider:
- Current MSE and target MSE
- Historical MSE trend (improving, declining, stagnant)
- Number of iterations completed
- Recent optimization attempts and their results
- Best practices for ML optimization

Provide a specific, actionable optimization strategy.`;

    const userPrompt = `Analyze this optimization scenario and recommend the best strategy:

CURRENT STATUS:
- Current MSE: ${analysisContext.currentMSE}
- Target MSE: ${analysisContext.targetMSE}
- Iteration: ${analysisContext.iterationNumber}
- Progress: ${analysisContext.context?.currentProgress || 'N/A'}

MSE TRENDS:
- Trend: ${analysisContext.trends.trend}
- Improving: ${analysisContext.trends.improvement}
- Stagnant: ${analysisContext.trends.stagnant}
- Recent MSEs: ${JSON.stringify(analysisContext.trends.recentMSEs)}

RECENT ATTEMPTS:
${JSON.stringify(analysisContext.recentAttempts.slice(-5), null, 2)}

Based on this analysis, recommend:
1. Primary strategy to apply
2. Specific technique or approach
3. Expected impact
4. Alternative strategy if primary fails
5. Whether to try a completely different approach

Format response as JSON with keys: strategy, technique, expectedImpact, alternativeStrategy, tryDifferentApproach (boolean)`;

    try {
      const message = await anthropic.messages.create({
        model: 'claude-3-5-sonnet-20241022',
        max_tokens: 1024,
        messages: [{
          role: 'user',
          content: userPrompt
        }],
        system: systemPrompt
      });

      const responseText = message.content[0].text;
      
      // Parse JSON response
      let jsonText = responseText;
      const jsonMatch = responseText.match(/```json\s*([\s\S]*?)\s*```/);
      if (jsonMatch) {
        jsonText = jsonMatch[1];
      } else {
        const codeMatch = responseText.match(/```\s*([\s\S]*?)\s*```/);
        if (codeMatch) {
          jsonText = codeMatch[1];
        }
      }
      
      return JSON.parse(jsonText);
    } catch (error) {
      return this._fallbackStrategy(
        analysisContext.mseHistory, 
        analysisContext.currentMSE, 
        analysisContext.targetMSE, 
        analysisContext.iterationNumber
      );
    }
  },
  
  /**
   * Fallback rule-based strategy when AI is not available
   */
  _fallbackStrategy(mseHistory, currentMSE, targetMSE, iterationNumber) {
    const trends = this._analyzeTrends(mseHistory);
    
    let strategy, technique;
    
    if (trends.stagnant) {
      strategy = 'Try different approach';
      technique = 'Switch to ensemble methods or different model architecture';
    } else if (!trends.improvement) {
      strategy = 'Simplify and regularize';
      technique = 'Add regularization, reduce model complexity';
    } else if (iterationNumber < 10) {
      strategy = 'Hyperparameter tuning';
      technique = 'Apply GridSearchCV for optimal parameters';
    } else if (iterationNumber < 20) {
      strategy = 'Feature engineering';
      technique = 'Add polynomial features and feature interactions';
    } else {
      strategy = 'Ensemble methods';
      technique = 'Combine multiple models with voting or stacking';
    }
    
    return {
      strategy,
      technique,
      expectedImpact: 'Moderate improvement expected',
      alternativeStrategy: 'Try cross-validation with different model',
      tryDifferentApproach: trends.stagnant || !trends.improvement
    };
  }
};

File: obscurex/tools/autonomy/recommend_optimization_strategy.py
"""
Tool: Recommend Optimization Strategy
Analyzes MSE trends and recommends optimization strategies
"""

from typing import Dict, Any, List


async def execute(params: Dict[str, Any], context: Dict[str, Any] = None) -> Dict[str, Any]:
    """Analyze MSE trends and recommend optimization strategy."""
    mse_history = params.get('mseHistory', [])
    current_mse = params.get('currentMSE')
    target_mse = params.get('targetMSE')
    iteration_number = params.get('iterationNumber', 0)
    ai_service = params.get('aiService')
    
    # Analyze trends
    trends = {
        'trend': 'unknown',
        'improvement': False,
        'stagnant': False,
        'avgImprovement': 0
    }
    
    if len(mse_history) >= 2:
        recent_improvements = [mse_history[i] - mse_history[i+1] for i in range(len(mse_history)-1)]
        avg_improvement = sum(recent_improvements) / len(recent_improvements) if recent_improvements else 0
        
        trends['avgImprovement'] = avg_improvement
        trends['improvement'] = avg_improvement > 0
        trends['trend'] = 'improving' if avg_improvement > 0 else 'declining'
        trends['stagnant'] = abs(avg_improvement) < 0.001
    
    # Get AI recommendation
    recommendation = {
        'strategy': 'Continue current approach',
        'technique': 'Keep optimizing with current methods',
        'expectedImpact': 'Moderate improvement expected',
        'tryDifferentApproach': trends.get('stagnant', False)
    }
    
    if ai_service and ai_service.is_available():
        try:
            prompt = f"""MSE History: {mse_history}
Current MSE: {current_mse}
Target MSE: {target_mse}
Iteration: {iteration_number}

Recommend an optimization strategy as JSON with keys: strategy, technique, expectedImpact, tryDifferentApproach."""
            
            ai_rec = await ai_service.analyze_and_decide(prompt, "You are an ML optimization expert.", 512)
            if isinstance(ai_rec, dict) and 'strategy' in ai_rec:
                recommendation = ai_rec
        except:
            pass
    
    return {
        'recommendation': recommendation,
        'trends': trends
    }


# Tool definition
tool = {
    'name': 'recommendOptimizationStrategy',
    'description': 'Analyze MSE trends and execution history to recommend the best optimization strategy',
    'parameters': {
        'mseHistory': 'array',
        'currentMSE': 'number',
        'targetMSE': 'number',
        'iterationNumber': 'number',
        'aiService': 'object|null'
    },
    'execute': execute
}
